103_agent-driven-attestation.md

```diff
@@ -0,0 +1,421 @@
+</p>
+
+If the outcome of all the trust decisions shown on the diagram is favourable, we can trust that the right policy (associated with a node's logical identifier) is applied to the right attestation (produced by a subordinate cryptographic identity) by following both chains all the way to a trusted root identity.
+
+Once again, we propose two different mechanism for making these trust decisions, which users can use independently or together at their discretion:
+
+- a standard trust store mechanism for use when certificates are available to establish trust; and/or
+- a webhook mechanism for when certificates are not an option or when custom trust logic is desirable.
+
+These are discussed in the [Design Details](#design-details) section.
+
+### Proposed Attestation Protocol
+
+To obtain an integrity quote in the current pull architecture, the verifier issues a request to the agent, supplying the following details: 
+
+- A nonce for the TPM to include in the quote
+- A mask indicating which PCRs should be included in the quote
+- An offset value indicating which IMA log entries should be sent by the agent
+
+The agent then replies with:
+
+- The UEFI measured boot log (kept in `/sys/kernel/security/tpm0/binary_bios_measurements`)
+- A list of IMA entries from the given offset
+- A quote of the relevant PCRs generated and signed by the TPM using the nonce
+
+In a push version of the protocol where the UEFI logs, IMA entries and quote are delivered to the verifier as an HTTP request issued by the agent, the agent needs a mechanism to first obtain the nonce, PCR mask and IMA offset from the verifier. We suggest simply adding a new HTTP endpoint to the verifier to make this information available to an agent correctly authenticated with the expected certificate via mTLS.
+
+As such, the push attestation protocol would operate in this manner:
+
+1. When it is time to report the next scheduled attestation, the agent will request the attestation details from the verifier.
+
+2. If the request is well formed, the verifier will reply with a new randomly-generated nonce and the PCR mask and IMA offset obtained from its database. Additionally, the verifier will persist the nonce to the database.
+
+3. The agent will gather the information required by the verifier (UEFI log, IMA entries and quote) and report these in a new HTTP request along with other information relevant to the quote (such as algorithms used).
+
+4. The verifier will reply with the number of seconds the agent should wait before performing the next attestation and an indication of whether the request from agent appeared well formed according to basic validation checks. Actual processing and verification of the measurements against policy can happen asynchronously after the response is returned. 
```

Comment:
If it is too soon, @mheese has suggested that the verifier could reply with a 429 status which I think makes a lot of sense—I just forgot to add it. I'll amend the proposal now to include that detail.

For the "too late" case: I suppose it would make sense to specify a period of validity for the nonce, after which it would expire. If an agent submits an attestation using a nonce that is no longer valid, the verifier would then reply with 400 status. What are your thoughts on this?

Reply:
The "too many requests" and the TTL for the nonce makes a lot of sense :+1:. We can make the TTL > TimeForNextAttestation, so we can play with network delays. This clarify the protocol.

What would be the consideration from the attestation PoV? If there is a 429 I guess that this will not make the note failed in the attestation state (it should be considered trusted), but maybe the 400 should require a change of status if it is repeated?

Reply:
To restate all the main error conditions we are considering here:

1. The verifier receives a request before it is expected, in which case it replies with a 429 (too many requests) and the number of seconds to wait in the `Retry-After` header.
2. The verifier receives an attestation for a nonce that is invalid or has expired, in which case it replies with a 400 (bad request).
3. The verifier receives an attestation which violates the policy for the node. In this case, the verifier will reject the next attestation received with a 503 (service unavailable). Of course, once this condition occurs, the verifier will begin reporting the status of the node as having failed attestation when queried.

In the first case, the agent should wait the number of seconds specified by the `Retry-After` header and try again.

In both the second and third case, the agent will retry according to an exponential backoff. This makes sense in the second case as that situation is unexpected behaviour and should only happen in practice, e.g., if the verifier is overloaded and taking a long time to respond to requests.

Exponential backoff is also appropriate in the third case as the user could change the verification policy at any time. As a result, the node cannot simply stop sending attestations when a verification failure occurs. Instead, it will continue to retry periodically in the hopes that an attestation will eventually be accepted.

Reply:
To reiterate, a 400 is returned when the verifier receives an attestation for a nonce that is invalid or has expired.

This may occur if the agent or node is behaving unexpectedly. However, it can also occur if the network or the verifier is behaving unexpectedly. E.g., if the verifier is overloading and taking a long time to reply to requests, this could delay the delivery of nonces and cause them to be used outside their period of validity.

So, we cannot use repeated 400s as a reliable indicator of a problem with the node. The verifier should only change the status of the node if an attestation fails verification or if no valid attestation has been received for the configured period of time.

Reply:
maybe to add on to that: a successful response should always include a timestamp (or a time duration) on when to query again, so that the agent knows when to perform another request. If it is too early, it receives a 429. (At least that was my idea when we discussed it with @stringlytyped )

Reply:
@mheese I agree that for 429s the verifier should include the seconds to wait before trying again (in the `Retry-After` header).

But for 400s, there is no way the verifier can know how long the agent should wait before trying again. Really, if everything is working fine, then the agent should be able to try again straight away by requesting a new nonce.

The 400 condition should not occur under normal operation. So, it makes sense for the agent to wait a while and then try again, increasing the wait time at each attempt until reaching a maximum wait time. This gives the verifier a chance to recover, if it happens that the condition has been caused by an overloaded verifier.